### ISSUE TITLE
How to generate QR from binary data?

### ISSUE THREAD

**Issue by Aleph72** [TheLevelUp/ZXingObjC#252]
I'm generating a QR code from a >500 characters string.
Now I've been asked to generate the QR from a "binary codification" to reduce its size and make it more readable.
Anyone knows what that means and how to do it? I have no idea...

Thanks.

**Comment by benjohnde** [comment]
Hi @Aleph72 you could generate a base64 string of this "binary data" and try that.

**Comment by Aleph72** [comment]
Hi @benjohnde, I don't have the binary data. All I have is a string with a lenght of more than 500 characters that generates a quite complex QR code.
My client asked me to pass the data as binary, instead of string, to the code generator.
If I convert my string to binary, or base64, or HEX, the result is a bigger string.
Is there any way to, for example, pass an NSData directly to the code generator? And will this reduce in any way the complexity of the generated QR code?

**Comment by benjohnde** [comment]
The magic happens here:

https://github.com/TheLevelUp/ZXingObjC/blob/master/ZXingObjC/qrcode/encoder/ZXQRCodeEncoder.m#L455

An option for passing NSData is currently not implemented. You would need to modify at least ZXQRCodeWriter.m and ZXQRCodeEncoder.m.

Another option, as long as you need to use QRCodes:

- https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html
- https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html#//apple_ref/doc/filter/ci/CIQRCodeGenerator
- http://www.appcoda.com/qr-code-generator-tutorial/

and finally more specifically: https://github.com/aschuch/QRCode/blob/master/QRCode/QRCode.swift

TL;DR: you could use core image library from iOS to create a qrcode by passing NSData natively.

**Comment by benjohnde** [comment]
@Aleph72 did it help?

**Comment by Aleph72** [comment]
Yes, a lot. Thank you very much.

**Comment by benjohnde** [comment]
You're welcome.
